File: backend/api/content_filter.py
"""
Sistema de filtragem de conteúdo sensível para proteger inputs do usuário.
"""
import re
from typing import Tuple


class ContentFilter:
    """
    Sistema de filtragem de conteúdo sensível para proteger inputs do usuário.
    """
    
    # Palavras e termos bloqueados (categorizados)
    BLOCKED_WORDS = {
        'sexual': [
            'sexo', 'pornografia', 'porno', 'xxx', 'nudez', 'nu', 'nua',
            'prostituição', 'prostituta', 'sexual', 'erótico', 'adulto', 'fetiche',
            'sexu', 'erotic', 'porn', 'putaria', 'puta', 'putas', 'puteiro',
            'safad', 'tesão', 'tesao', 'gostosa', 'gostoso', 'bundão', 'bundao',
            'peito', 'seios', 'vagina', 'pênis', 'penis', 'genital', 'orgia',
            'masturbação', 'masturba', 'transa', 'sexting', 'nudes', 'pelad',
            'stripper', 'prostitui', 'escort', 'sensual', 'provocante'
        ],
        'violence': [
            'matar', 'assassinar', 'suicídio', 'violência', 'tortura',
            'agressão', 'espancar', 'sangue', 'morte', 'arma', 'faca',
            'tiro', 'bomba', 'explosivo', 'terrorismo', 'terrorista',
            'estupro', 'violentar', 'abusar', 'massacre', 'mutil',
            'enforcar', 'guilhotin', 'decapit', 'esfaquear', 'atirar'
        ],
        'hate': [
            'racismo', 'nazi', 'nazismo', 'homofobia', 'xenofobia', 'preconceito',
            'discriminação', 'ódio', 'linchamento', 'fascismo', 'supremacia',
            'inferioridade', 'raça inferior', 'viado', 'bicha', 'sapatão',
            'macaco', 'preto', 'judeu', 'muçulman', 'terrorista'
        ],
        'illegal': [
            'drogas', 'tráfico', 'hacker', 'hackear', 'pirataria',
            'roubar', 'fraudar', 'golpe', 'crime', 'criminoso',
            'cocaína', 'maconha', 'heroína', 'crack', 'ecstasy',
            'traficar', 'contrabando', 'falsificação', 'estelionato'
        ],
        'injection': [
            '<script>', 'javascript:', 'onerror=', 'onclick=', 'onload=',
            'eval(', 'exec(', 'system(', 'DROP TABLE', 'DELETE FROM',
            'INSERT INTO', 'UPDATE SET', '--', ';--', '/*', '*/',
            '../', '..\\', 'file://', 'data:', 'vbscript:', 'base64'
        ]
    }
    
    # Padrões regex mais inteligentes para detectar variações
    BLOCKED_PATTERNS = [
        # Sexual - detecta variações de palavras relacionadas a sexo
        r'\b(?:sex|porn|xxx|puta|putar|safad|tesã|tesao|erótic|erotic)\w*',
        r'\b(?:prostitui|escort|strip|nudez|pelad)\w*',
        r'\b(?:masturb|transa|orgia|sensual)\w*',
        
        # Violência - detecta variações (MAIS ESPECÍFICO para não pegar "Matemática")
        r'\b(?:matar|matou|matei|matando|assassin|suicid|violent|tortur)\w*',
        r'\b(?:estupra|violenta|abusa|massacr|mutil)\w*',
        
        # Ódio - detecta termos ofensivos
        r'\b(?:racis|nazi|fascis|homofob|xenofob)\w*',
        r'\b(?:viado|bicha|sapatão|macaco)\b',
        
        # Drogas
        r'\b(?:cocaín|maconha|heroín|crack|ecstasy|trafic)\w*',
    ]
    
    # Padrões regex suspeitos
    SUSPICIOUS_PATTERNS = [
        r'<[^>]*script[^>]*>',  # Tags script
        r'javascript\s*:',       # Links javascript
        r'on\w+\s*=',           # Event handlers HTML
        r'eval\s*\(',           # Funções perigosas
        r'exec\s*\(',
        r'DROP\s+TABLE',        # SQL injection
        r'DELETE\s+FROM',
        r'INSERT\s+INTO',
        r'\.\./|\.\.\\',        # Path traversal
        r'file\s*:\s*//',       # File protocol
        r'data\s*:\s*text',     # Data URLs
    ]
    
    @classmethod
    def is_safe(cls, text: str) -> Tuple[bool, str]:
        """
        Verifica se o texto é seguro.
        
        Returns:
            Tuple[bool, str]: (is_safe, reason_if_unsafe)
        """
        if not text:
            return True, ""
        
        text_lower = text.lower().strip()
        
        # 1. Verifica padrões regex (detecta variações)
        for pattern in cls.BLOCKED_PATTERNS:
            if re.search(pattern, text_lower, re.IGNORECASE):
                return False, "Conteúdo inapropriado detectado"
        
        # 2. Verifica palavras bloqueadas exatas
        for category, words in cls.BLOCKED_WORDS.items():
            for word in words:
                # Verifica palavra exata ou como parte de palavra
                if word in text_lower:
                    return False, "Conteúdo inapropriado detectado"
        
        # 3. Verifica padrões suspeitos (injection, XSS, etc)
        for pattern in cls.SUSPICIOUS_PATTERNS:
            if re.search(pattern, text, re.IGNORECASE):
                return False, "Padrão de código malicioso detectado"
        
        # 4. Verifica tamanho excessivo (possível ataque)
        if len(text) > 500:
            return False, "Texto muito longo (máximo 500 caracteres)"
        
        # 5. Verifica caracteres especiais excessivos
        special_chars = sum(1 for c in text if not c.isalnum() and not c.isspace() and c not in 'áàâãéèêíïóôõöúçñÁÀÂÃÉÈÊÍÏÓÔÕÖÚÇÑ')
        if special_chars > len(text) * 0.3:  # Mais de 30% de caracteres especiais
            return False, "Muitos caracteres especiais"
        
        return True, ""
    
    @classmethod
    def sanitize(cls, text: str) -> str:
        """
        Limpa o texto removendo caracteres perigosos.
        """
        if not text:
            return ""
        
        # Remove tags HTML
        text = re.sub(r'<[^>]+>', '', text)
        
        # Remove caracteres de controle
        text = re.sub(r'[\x00-\x1F\x7F]', '', text)
        
        # Remove múltiplos espaços
        text = re.sub(r'\s+', ' ', text)
        
        return text.strip()

File: Frontend/src/utils/contentFilter.ts
/**
 * Sistema de filtragem de conteúdo sensível para proteger inputs do usuário
 */

export class ContentFilter {
  private static readonly BLOCKED_KEYWORDS = [
    // Conteúdo sexual - detecta variações e contexto
    /\b(sex|porn|xxx|puta|putar|putaria|safad|tesã|tesao|er[oó]tic|adulto|fetiche)\w*/i,
    /\b(prostitui|escort|strip|nudez|pelad|masturb|transa|orgia|sensual)\w*/i,
    /\b(gostosa|gostoso|bundão|bundao|peito|seios|vagina|p[eê]nis|genital)\w*/i,
    /\b(nudes|sexting|provocante)\w*/i,
    
    // Violência - detecta variações (MAIS ESPECÍFICO para não pegar "Matemática")
    /\b(matar|matou|matei|matando|assassi|suic[ií]d|violent|tortur|estupra|violenta|abusa|massacr|mutil)\w*/i,
    /\b(arma|bomba|terroris|sangue|morte|agressão|espancar)\w*/i,
    
    // Ódio - detecta termos ofensivos e suas variações
    /\b(racis|nazi|fascis|homofob|xenofob|discrimina[cç]ão|[oó]dio|linchamento)\w*/i,
    /\b(viado|bicha|sapatão|macaco|preto|judeu|mu[cç]ulman)\b/i,
    
    // Ilegal - drogas e crimes
    /\b(droga|tr[aá]fico|cocaín|maconha|hero[ií]n|crack|ecstasy|contrabando)\w*/i,
    /\b(hackear|pirataria|roubar|frauda|golpe|crime|estelionato)\w*/i,
    
    // Injection patterns
    /<script|javascript:|onerror=|onclick=|eval\(|DROP\s+TABLE|DELETE\s+FROM/i,
    /\.\.[/\\]|file:\/\/|data:text|base64/i,
  ];

  /**
   * Verifica se o texto é seguro
   */
  static isSafe(text: string): { safe: boolean; reason?: string } {
    if (!text || text.trim().length === 0) {
      return { safe: true };
    }

    const trimmedText = text.trim();

    // 1. Verifica tamanho
    if (trimmedText.length > 500) {
      return { safe: false, reason: "Texto muito longo (máximo 500 caracteres)" };
    }

    // 2. Verifica palavras bloqueadas
    for (const pattern of this.BLOCKED_KEYWORDS) {
      if (pattern.test(trimmedText)) {
        return { safe: false, reason: "Conteúdo inapropriado detectado" };
      }
    }

    // 3. Verifica caracteres especiais excessivos
    const allowedCharsPattern = /[a-zA-Z0-9\sáàâãéèêíïóôõöúçñÁÀÂÃÉÈÊÍÏÓÔÕÖÚÇÑ]/g;
    const allowedChars = trimmedText.match(allowedCharsPattern) || [];
    const specialChars = trimmedText.length - allowedChars.join('').length;
    
    if (specialChars > trimmedText.length * 0.3) {
      return { safe: false, reason: "Muitos caracteres especiais" };
    }

    // 4. Verifica caracteres repetidos (possível spam)
    if (/(.)\1{10,}/.test(trimmedText)) {
      return { safe: false, reason: "Padrão de spam detectado" };
    }

    return { safe: true };
  }

  /**
   * Sanitiza o texto removendo caracteres perigosos
   */
  static sanitize(text: string): string {
    if (!text) return "";

    return text
      // Remove tags HTML
      .replace(/<[^>]*>/g, '')
      // Remove caracteres de controle
      .replace(/[\x00-\x1F\x7F]/g, '')
      // Remove múltiplos espaços
      .replace(/\s+/g, ' ')
      .trim();
  }
}
